File: historyscan-back/src/main/java/io/sch/historyscan/infrastructure/features/codebase/CodeBaseApplication.java
package io.sch.historyscan.infrastructure.features.codebase;

import io.sch.historyscan.domain.contexts.codebase.clone.Clone;
import io.sch.historyscan.domain.contexts.codebase.clone.CodeBaseToClone;
import io.sch.historyscan.domain.contexts.codebase.find.FindCodeBase;
import io.sch.historyscan.domain.contexts.codebase.find.FindCodeBases;
import io.sch.historyscan.infrastructure.features.codebase.clone.AddedCodebaseDTO;
import io.sch.historyscan.infrastructure.features.codebase.clone.CodeBaseToAddDTO;
import io.sch.historyscan.infrastructure.features.codebase.exceptions.CodeBaseAlreadyExistsException;
import io.sch.historyscan.infrastructure.features.codebase.info.CodebaseDTO;
import io.sch.historyscan.infrastructure.features.codebase.list.CurrentCodebasesDTO;
import io.sch.historyscan.infrastructure.hexagonalarchitecture.HexagonalArchitectureApplication;
import org.springframework.stereotype.Component;

import java.util.Optional;

@Component
@HexagonalArchitectureApplication
public class CodeBaseApplication {

    private final Clone clone;
    private final FindCodeBases findCodeBases;
    private final FindCodeBase findCodeBase;
    private final CodeBaseMapper codeBaseMapper;

    public CodeBaseApplication(Clone clone,
                               FindCodeBases findCodeBases,
                               FindCodeBase findCodeBase, CodeBaseMapper codeBaseMapper) {
        this.clone = clone;
        this.findCodeBases = findCodeBases;
        this.findCodeBase = findCodeBase;
        this.codeBaseMapper = codeBaseMapper;
    }

    public AddedCodebaseDTO clone(CodeBaseToAddDTO codeBaseToAddDTO) throws CodeBaseAlreadyExistsException {
        var codebase = clone.from(new CodeBaseToClone(
                codeBaseToAddDTO.url(), codeBaseToAddDTO.publicKey(),
                codeBaseToAddDTO.name(), codeBaseToAddDTO.branch()));
        return codeBaseMapper.domainToWeb(codebase);

    }

    public Optional<CodebaseDTO> findCodeBase(String codebaseName) {
        return findCodeBase.from(codebaseName)
                .map(codeBaseMapper::codebaseDomainToWeb);
    }

    public CurrentCodebasesDTO currentCodeBases() {
        var codebases = findCodeBases.fromDisk();
        return codeBaseMapper.domainToWeb(codebases);
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/infrastructure/features/codebase/CodeBaseApplication.kt
package io.sch.historyscan.infrastructure.features.codebase

import io.sch.historyscan.domain.contexts.codebase.clone.*
import io.sch.historyscan.domain.contexts.codebase.find.FindCodeBase
import io.sch.historyscan.domain.contexts.codebase.find.FindCodeBases
import io.sch.historyscan.infrastructure.features.codebase.clone.AddedCodebaseDTO
import io.sch.historyscan.infrastructure.features.codebase.clone.CodeBaseToAddDTO
import io.sch.historyscan.infrastructure.features.codebase.info.CodebaseDTO
import io.sch.historyscan.infrastructure.features.codebase.list.CurrentCodebasesDTO
import io.sch.historyscan.infrastructure.hexagonalarchitecture.HexagonalArchitectureApplication
import org.springframework.stereotype.Component
import java.util.*

@Component
@HexagonalArchitectureApplication
class CodeBaseApplication(
    private val clone: Clone,
    private val findCodeBases: FindCodeBases,
    private val findCodeBase: FindCodeBase, private val codeBaseMapper: CodeBaseMapper
) {
    @Throws(CodeBaseAlreadyExistsException::class)
    fun clone(codeBaseToAddDTO: CodeBaseToAddDTO): AddedCodebaseDTO {
        val codebase = clone.from(
            CodeBaseToClone(
                codeBaseToAddDTO.url, codeBaseToAddDTO.publicKey,
                codeBaseToAddDTO.name, codeBaseToAddDTO.branch
            )
        )
        return codeBaseMapper.domainToWeb(codebase)
    }

    fun findCodeBase(codebaseName: String): CodebaseDTO? {
        return findCodeBase.from(codebaseName)?.let { codeBaseMapper.codebaseDomainToWeb(it) }
    }

    fun currentCodeBases(): CurrentCodebasesDTO? {
        val codebases = findCodeBases.fromDisk()
        return codeBaseMapper.domainToWeb(codebases)
    }
}
